src/include/processor/operator/aggregate/simple_aggregate.h

```diff
@@ -8,23 +10,68 @@ namespace processor {
 
 // NOLINTNEXTLINE(cppcoreguidelines-virtual-class-destructor): This is a final class.
 class SimpleAggregateSharedState final : public BaseAggregateSharedState {
+    friend class SimpleAggregate;
+
 public:
-    explicit SimpleAggregateSharedState(
-        const std::vector<function::AggregateFunction>& aggregateFunctions);
+    explicit SimpleAggregateSharedState(main::ClientContext* clientContext,
+        const std::vector<function::AggregateFunction>& aggregateFunctions,
+        const std::vector<AggregateInfo>& aggInfos);
+
+    // The partitioningData objects need a stable pointer to this shared state
+    DELETE_COPY_AND_MOVE(SimpleAggregateSharedState);
 
     void combineAggregateStates(
         const std::vector<std::unique_ptr<function::AggregateState>>& localAggregateStates,
         storage::MemoryManager* memoryManager);
 
-    void finalizeAggregateStates();
+    void finalizeAggregateStates(storage::MemoryManager* memoryManager);
 
     std::pair<uint64_t, uint64_t> getNextRangeToRead() override;
 
     function::AggregateState* getAggregateState(uint64_t idx) {
         return globalAggregateStates[idx].get();
     }
 
+    // Merges data from the queues into the distinct hash tables
+    // Can be run concurrently (but only after all data has been written into the queues)
+    void finalizeDistinctHashTables(storage::MemoryManager* memoryManager,
+        const std::vector<AggregateInfo>& aggInfos);
+
+    bool isReadyForFinalization() const { return readyForFinalization; }
+
+protected:
+    struct Partition {
+        struct DistinctData {
+            std::unique_ptr<AggregateHashTable> hashTable;
+            std::unique_ptr<HashTableQueue> queue;
+            std::unique_ptr<function::AggregateState> state;
+        };
+        std::mutex mtx;
+        std::vector<DistinctData> distinctTables;
+        std::atomic<bool> finalized = false;
+    };
+
+    class SimpleAggregatePartitioningData : public AggregatePartitioningData {
+    public:
+        SimpleAggregatePartitioningData(SimpleAggregateSharedState* sharedState, size_t functionIdx)
+            : sharedState{sharedState}, functionIdx{functionIdx} {}
+
+        void appendTuples(const FactorizedTable& factorizedTable,
+            ft_col_offset_t hashOffset) override;
+        void appendDistinctTuple(size_t, std::span<uint8_t>, common::hash_t) override;
+        void appendOverflow(common::InMemOverflowBuffer&& overflowBuffer) override;
+
+    private:
+        SimpleAggregateSharedState* sharedState;
+        size_t functionIdx;
+    };
+
 private:
+    bool readyForFinalization;
+    bool hasDistinct;
+    common::MPSCQueue<std::unique_ptr<common::InMemOverflowBuffer>> overflow;
+    std::vector<Partition> globalPartitions;
```

Comment:
`readyForFinalization`, `overflow` and `globalPartitions`  can be moved into `BaseAggregateSharedState`.

Reply:
globalPartitions can't because there is a different Partition type for each of the two derived classes, but the others certainly can.